feat: [M3-8856] - Surface Labels and Taints for LKE node pools - Part 2

mjac0bs opened the pull request:
## Description 📝

See #11528 for Part 1. This PR, Part 2, will include support for adding Labels and Taints to the node pool.

WIP - Putting this up for early UX feedback.


## Changes  🔄

List any change(s) relevant to the reviewer.

- ...
- ...

## Target release date 🗓️

Please specify a release date (and environment, if applicable) to guarantee timely review of this PR. If exact date is not known, please approximate and update it as needed.

## Preview 📷

**Include a screenshot or screen recording of the change.**

:lock: Use the [Mask Sensitive Data](https://cloud.linode.com/profile/settings) setting for security.

:bulb: Use `<video src="" />` tag when including recordings in table.

| Before  | After   |
| ------- | ------- |
| 📷 | 📷 |

## How to test 🧪

### Prerequisites

(How to setup test environment)

- ...
- ...

### Reproduction steps

(How to reproduce the issue, if applicable)

- [ ] ...
- [ ] ...

### Verification steps

(How to verify changes)

- [ ] ...
- [ ] ...

<details>
<summary> Author Checklists </summary>

## As an Author, to speed up the review process, I considered 🤔

👀 Doing a self review
❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
🤏 Splitting feature into small PRs
➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
🧪 Providing/improving test coverage
 🔐 Removing all sensitive information from the code and PR description
🚩 Using a feature flag to protect the release
👣 Providing comprehensive reproduction steps
📑 Providing or updating our documentation
🕛 Scheduling a pair reviewing session
📱 Providing mobile support
♿  Providing accessibility support

<br/>

- [ ] I have read and considered all applicable items listed above.

## As an Author, before moving this PR from Draft to Open, I confirmed ✅

- [ ] All unit tests are passing
- [ ] TypeScript compilation succeeded without errors
- [ ] Code passes all linting rules

</details>

mjac0bs commented on packages/manager/cypress/e2e/core/kubernetes/lke-update.spec.ts:
It's probably easier to just read through this spec locally (`confirms labels and taints functionality for a node pool`). In this PR, I added the new test for `can add labels and taints`, and moved some stuff around so `can delete labels and taints` can share some of the same set up.

mjac0bs commented on packages/manager/src/features/Kubernetes/KubernetesClusterDetail/NodePoolsDisplay/LabelsAndTaints/LabelInput.tsx:
```diff
@@ -0,0 +1,96 @@
+
+    try {
+      clearErrors();
+      kubernetesLabelSchema.validateSync(newLabels);
```
We are validating here rather than when the form is submitted because it seems more intuitive to give the user immediate feedback when they're using the add form, rather than when all changes are being saved.

mjac0bs commented on packages/manager/src/features/Kubernetes/KubernetesClusterDetail/NodePoolsDisplay/LabelsAndTaints/TaintTable.tsx:
```diff
@@ -15,10 +17,13 @@ export const TaintTable = () => {
     setValue(
       'taints',
-      taints.filter((taint) => taint.key !== key),
+      taints.filter(
+        (taint) =>
+          taint.key !== removedTaint.key || taint.value !== removedTaint.value
+      ),
```
Fixes a bug where, if multiple taints had the same key, deleting one of those taints would delete the rest of them with that key.

mjac0bs replied:
Similar - thanks for raising. Taints have the same key-value, but different effects. I'm going to verify API behavior to see if it even allows taints with the same key-value but differing effects.

mjac0bs replied:
Fixed the deletion in [bd3dd94](https://github.com/linode/manager/pull/11553/commits/bd3dd943c17ed65c2d5a890d96a2cc7bda457d07), because the API does accept multiple identical key-taint pairs as valid.


Though: I'm looking into this a bit further, because I don't think they will actually get *applied* if that's the case. 

[kubectl docs](https://kubernetes.io/docs/reference/kubectl/generated/kubectl_taint/#examples) suggest that the `value` will actually get updated:
```
# Update node 'foo' with a taint with key 'dedicated' and value 'special-user' and effect 'NoSchedule'
  # If a taint with that key and effect already exists, its value is replaced as specified
  kubectl taint nodes foo dedicated=special-user:NoSchedule
```

mjac0bs commented on packages/validation/src/kubernetes.schema.ts:
```diff
@@ -88,3 +88,83 @@ const controlPlaneACLOptionsSchema = object().shape({
 });
+
+// Starts with a letter or number and contains letters, numbers, hyphens, dots, and underscores
+const alphaNumericValidCharactersRegex = /^[a-zA-Z0-9]([a-zA-Z0-9-._]*[a-zA-Z0-9])?$/;
+
+// DNS subdomain prefix (example.com/my-app)
+const dnsPrefixRegex = /^[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?(\.[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?)*$/;
```
I should add better test coverage to check this validation before I open this up for review. It's not pretty.

mjac0bs commented on packages/validation/src/kubernetes.schema.ts:
```diff
@@ -88,3 +88,83 @@ const controlPlaneACLOptionsSchema = object().shape({
+const MAX_DNS_KEY_SUFFIX_LENGTH = 62;
+const MAX_SIMPLE_KEY_OR_VALUE_LENGTH = 63;
+
+const validateKubernetesLabel = (labels: {
```
See https://techdocs.akamai.com/cloud-computing/docs/deploy-and-manage-a-kubernetes-cluster-with-the-api#add-labels-and-taints-to-your-lke-node-pools for the validation rules.

I'm open to any ideas on making this validation util less complicated.

hana-akamai replied:
Can we let the API handle complicated validation?

mjac0bs replied:
The user would have to try to save changes on the whole form first, so feedback would be less immediate. The user would then have to delete the invalid label from the table and re-add the correct version, which is more steps. I'll take a look, though.

mjac0bs replied:
@hana-akamai After looking into this, I'm thinking we should attempt to client-side validate (even if the schema is complex) when possible here for a better UX, and fall back on API field errors (surfaced correctly now [5324e08](https://github.com/linode/manager/pull/11553/commits/5324e0819366fd0f9e75406a4202fa543050f4b1)) in if there's anything our validation misses.

API sometimes returns specific but misleading error messages, which seems worse than a more generic message. Some examples I ran into with labels, when relying solely on the API for validation:

| `labels` | Header | Header |
|--------|--------|--------|
| `{ 'label1': 'label 1', 'label2': 'label 2'}` | Mentions nothing about spaces | ![Screenshot 2025-01-31 at 8 33 44 AM](https://github.com/user-attachments/assets/7cdf58cf-3d58-4f1e-b4e6-e1e2a5bdaef0) |
| `{ 'fMThq396QSnujcX8YEX75b4TKQaG3GvT7yK7NV15eWaLV3fkbK0Ng0QDQeYMaWFpc2BYbMPdfvDSZh8KfJkxcCbwkxNcidYHDWvp09ieZg54vPtkJ2dcER6Ld696VMBUhNctNvHu2fPGH2DWWDAc2C17AT7DUUM8SF7Umx6FBg8tnbDaNqHi7FpZHDq1e6WgVJPmYRQLB1Q3JAaZeHu7EK7bp9mDZejV5xey3gQiFWWCJyVmQmyDjyrjCte3YAkKGPeQYbbwkqWgaDEyjzQeM1Dm.io/test: test': 'test}` | Mentions nothing about invalid length | ![Screenshot 2025-01-31 at 8 35 48 AM](https://github.com/user-attachments/assets/83fea064-c19d-423d-8008-6ee36754e1fd) |
| `{ '': ''}` | We'd still need to validate on the empty field | <video src="https://github.com/user-attachments/assets/1fb86458-7fbb-4998-b28d-7887347fb05b" /> | 

It also looks like we'd have to work to handle surfacing more than one API error at a time -- we don't have form fields for each field error, just the table rows. I'd hoped we could do something like [this](packages/manager/src/features/Kubernetes/KubernetesClusterDetail/NodePoolsDisplay/LabelsAndTaints/TaintTable.tsx) to surface error rows on the table, but ran into issues with getting the errors to persist as expected (and we'd still be surfacing some misleading error messages).

https://github.com/user-attachments/assets/31cbb6fa-5b55-4b56-8b86-9941e5e45e59

hana-akamai reviewed (commented):
Something seems off with the Taint validation, seeing this error when trying to input the placeholder

Can we also make Label and Taints form hidden again on cancel?

https://github.com/user-attachments/assets/4e3cbd47-37fb-4833-b424-8c9da5bb8036

mjac0bs commented:
@hana-akamai 

> Something seems off with the Taint validation, seeing this error when trying to input the placeholder

Thanks for catching. This was happening because the `dnsPrefixRegex` didn't include the slash as a valid character and because `matches` can't take an or operator - not sure why I thought (or hoped) it did. Updated the regex and switched to a `test` function in  [cbfb49d](https://github.com/linode/manager/pull/11553/commits/cbfb49db5b0fcb84bcd134e943a24988937f0dbe). I also updated the Cypress test coverage to test both adding labels and taints with a simple key and with a DNS subdomain in the key, which would have caught this.

> Can we also make Label and Taints form hidden again on cancel?

Sure, did that in [cfe0979](https://github.com/linode/manager/pull/11553/commits/cfe09794e4dcab7902fa178c4e579a5480ad1d09).

mjac0bs commented on packages/manager/src/features/Kubernetes/KubernetesClusterDetail/NodePoolsDisplay/NodePool.tsx:
```diff
@@ -123,14 +119,12 @@ export const NodePool = (props: Props) => {
+              compactY
+              onClick={() => handleClickLabelsAndTaints(poolId)}
+            >
+              Labels and Taints
```
It does, but this is app-wide and I'd say beyond the scope of this PR.

mjac0bs commented on packages/manager/src/features/Kubernetes/KubernetesClusterDetail/NodePoolsDisplay/LabelsAndTaints/TaintTable.tsx:
```diff
@@ -15,10 +17,15 @@ export const TaintTable = () => {
     setValue(
       'taints',
-      taints.filter((taint) => taint.key !== key),
+      taints.filter(
+        (taint) =>
+          taint.key !== removedTaint.key ||
+          taint.value !== removedTaint.value ||
+          taint.effect !== removedTaint.effect
+      ),
```
@dwiley-akamai Not sure why my other comment isn't showing up. I fixed this so that deleting a row will only delete the taint with the given key, value, and effect.

Not related to this bug directly, but I did some confirming of what taints should actually be valid in this table:
API does accept a taint with the same key-value and different effects. I'm not certain how that actually gets applied to pods (but think kubernetes should worry about that, as long as API accepts it, for parity). The API does enforce uniqueness of key-effect pairs and we surface that in an API message. I'd be okay with letting API validation handle that edge case, rather than try to handle it ourselves.

<details>
<summary>Screenshots</summary>

![Screenshot 2025-01-31 at 9 30 12 AM](https://github.com/user-attachments/assets/e0f30edf-b341-4194-8a20-07a2c071ee9e)
![Screenshot 2025-01-31 at 9 28 31 AM](https://github.com/user-attachments/assets/949f3c19-1458-4045-96d5-431ef895b43c)

</details>

hana-akamai replied:
Does the API enforce uniqueness of key-value labels? (Right now it's not possible to have the same key in the UI, it gets overridden)

<details>
<summary> Video </summary>

https://github.com/user-attachments/assets/2f2a40b7-c99d-4bb7-8858-bca729ca4874


</details>

mjac0bs replied:
Yes. Because `labels` is an object and JS objects can only have unique keys. 

And here's the [docs](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/) reference: 
![image](https://github.com/user-attachments/assets/5237ae48-836f-4337-a924-6f67eea2dab7)

This *is* behaving as expected for Labels.

mjac0bs reviewed (commented):
This should be ready for re-review after a look into some error handling and addressing the bugs mentioned previously.

Thanks @dwiley-akamai and @hana-akamai - I appreciate all your feedback!

hana-akamai reviewed (commented):
Can we improve keyboard navigation? Right now, when deleting an item, we lose focus and have to navigate back to the next item. The cancel button is also the button that's focused instead of Add and I wasn't able to navigate to Add

<details>
<summary> Videos </summary>

| Labels and Taints  | MultipleIP for reference   |
| ------- | ------- |
| <video src="https://github.com/user-attachments/assets/b6783e1e-072f-4fc6-9c2c-16c60c7814dd" /> | <video src="https://github.com/user-attachments/assets/0bb5b27c-9553-4ed3-8966-b3c033b15c7e" /> |

</details>

dwiley-akamai reviewed (approved):
Ability to add, save, and edit Labels and Taints ✅
Validation for `Label` and `Taint` fields ✅
`lke-update.spec.ts` passes locally & in CI ✅
No regressions observed for LKE pools ✅
Code review ✅

hana-akamai reviewed (approved):
Nice work! 🚀
